tests/integration/models/device.spec.coffee

```diff
@@ -890,3 +890,220 @@ describe 'Device Model', ->
+						resin.models.device.tags.getAllByDevice(@device.id)
+					.then (tags) ->
+						m.chai.expect(tags).to.have.length(1)
+						m.chai.expect(tags[0].tag_key).to.equal('EDITOR')
+						m.chai.expect(tags[0].value).to.equal('vim')
+
+				it 'should be able to create a numeric tag', ->
+					resin.models.device.tags.set(@device.uuid, 'EDITOR', 1).then =>
+						resin.models.device.tags.getAllByDevice(@device.id)
+					.then (tags) ->
+						m.chai.expect(tags).to.have.length(1)
+						m.chai.expect(tags[0].tag_key).to.equal('EDITOR')
+						m.chai.expect(tags[0].value).to.equal('1')
+
+				it 'should not allow creating a resin tag', ->
+					promise = resin.models.device.tags.set(@device.uuid, 'io.resin.test', 'secret')
+					m.chai.expect(promise).to.be.rejectedWith('Tag keys beginning with io.resin. are reserved.')
+
+				it 'should not allow creating a tag with a name that starts with a number', ->
+					promise = resin.models.device.tags.set(@device.uuid, '9_EDITOR', 'vim')
+					m.chai.expect(promise).to.be.rejectedWith(/Invalid parameter: 9_EDITOR is not a valid value/)
+
+				it 'should not allow creating a tag with a name containing an illegal character', ->
+					promise = resin.models.device.tags.set(@device.uuid, 'EDITOR=', 'vim')
+					m.chai.expect(promise).to.be.rejectedWith(/Invalid parameter: EDITOR= is not a valid value/)
+
+				it 'should be rejected if the device uuid does not exist', ->
+					promise = resin.models.device.tags.set('asdfghjkl', 'EDITOR', 'vim')
+					m.chai.expect(promise).to.be.rejectedWith('Device not found: asdfghjkl')
+
+				it 'should be rejected if the device id does not exist', ->
+					promise = resin.models.device.tags.set(999999, 'EDITOR', 'vim')
+					m.chai.expect(promise).to.be.rejectedWith('Device not found: 999999')
+
+				describe 'given two existing tags', ->
+
+					beforeEach ->
+						Promise.all([
+							resin.models.device.tags.set(@device.id, 'EDITOR', 'vim')
+							resin.models.device.tags.set(@device.id, 'LANGUAGE', 'js')
+						])
+						.then (args) =>
+							[@tag, @tag2] = args
+
+					it 'should be able to update a tag without affecting the rest', ->
+						resin.models.device.tags.set(@device.id, 'EDITOR', 'emacs')
+						.then => resin.models.device.tags.getAllByDevice(@device.id)
+						.then (tags) ->
+							tags = _.sortBy(tags, 'tag_key')
+							m.chai.expect(tags).to.have.length(2)
+							m.chai.expect(tags[0].tag_key).to.equal('EDITOR')
+							m.chai.expect(tags[0].value).to.equal('emacs')
+							m.chai.expect(tags[1].tag_key).to.equal('LANGUAGE')
+							m.chai.expect(tags[1].value).to.equal('js')
+
+					afterEach ->
+						resin.models.device.tags.remove(@device.id, 'LANGUAGE')
+
+			describe 'given an existing tag', ->
+
+				beforeEach ->
+					resin.models.device.tags.set(@device.id, 'EDITOR', 'vim').then (tag) =>
+						@tag = tag
+
+				describe 'resin.models.device.tags.remove()', ->
```

Comment:
In all these tests, there's a lot of duplication. Can we reduce that, to simplify & shorten these a bit?

Doesn't need to be one function for the whole lot, especially when there's clearly awkward special cases - we can have a few methods like `itShouldAllowRemovable(resin.models.device.tags)` or something, to run the `.remove` tests for the various cases, and then end up just composing a few reusable functions like that together, plus the extra cases, to get this same test suite?

It's pretty noisy already, and it's only going to get worse once we add in release tags too. I'm not totally sure what this would look like, but it'd be useful if you could play around a little here and see if there are any simple routes to improve it.